cmd/sonobuoy/app/images.go

```diff
@@ -22,46 +22,102 @@ import (
 
 	"github.com/heptio/sonobuoy/pkg/errlog"
 	"github.com/heptio/sonobuoy/pkg/image"
+	"github.com/heptio/sonobuoy/pkg/image/docker"
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
-	"github.com/spf13/pflag"
 )
 
 var imagesflags imagesFlags
 
-type imagesFlags struct {
-	e2eRegistryConfig string
-	plugin            string
-	kubeconfig        Kubeconfig
-
-	imagesflags *pflag.FlagSet
-}
+const numDockerRetries = 1
+)
 
-func ImagesFlagSet(cfg *imagesFlags) *pflag.FlagSet {
-	flagset := pflag.NewFlagSet("images", pflag.ExitOnError)
-	AddPluginFlag(&cfg.plugin, flagset)
-	AddKubeconfigFlag(&cfg.kubeconfig, flagset)
-	cfg.imagesflags = flagset
-	return flagset
+type imagesFlags struct {
+	e2eRegistryConfig  string
+	plugin             string
+	username           string
+	password           string
+	imagesSaveFileName string
+	downloadTar        bool
+	kubeconfig         Kubeconfig
 }
 
 func NewCmdImages() *cobra.Command {
+	// Main command
 	cmd := &cobra.Command{
 		Use:   "images",
 		Short: "Manage images used in a plugin. Supported plugins are: 'e2e'",
-		Run:   getImages,
+		Run:   listImages,
 		Args:  cobra.ExactArgs(0),
 	}
 
-	cmd.Flags().AddFlagSet(ImagesFlagSet(&imagesflags))
+	AddKubeconfigFlag(&imagesflags.kubeconfig, cmd.Flags())
+	AddPluginFlag(&imagesflags.plugin, cmd.Flags())
+
+	// Pull command
+	pullCmd := &cobra.Command{
+		Use:   "pull",
+		Short: "Pulls images to local docker client for a specific plugin",
+		Run:   pullImages,
+		Args:  cobra.ExactArgs(0),
+	}
+	AddKubeconfigFlag(&imagesflags.kubeconfig, pullCmd.Flags())
+	AddPluginFlag(&imagesflags.plugin, pullCmd.Flags())
+
+	// Download command
+	downloadCmd := &cobra.Command{
+		Use:   "download",
+		Short: "Saves downloaded images from local docker client to a tar file",
+		Run:   downloadImages,
+		Args:  cobra.ExactArgs(0),
+	}
+	AddKubeconfigFlag(&imagesflags.kubeconfig, downloadCmd.Flags())
+	AddPluginFlag(&imagesflags.plugin, downloadCmd.Flags())
+
+	// Push command
+	pushCmd := &cobra.Command{
+		Use:   "push",
+		Short: "Pushes images to docker registry for a specific plugin",
+		Run:   pushImages,
+		Args:  cobra.ExactArgs(0),
```

Comment:
Just an idea (and maybe this will just be a future ticket) but maybe in the future we allow N args to these commands so that users can push/pull just the list of tests they want? Either just the ones from a given registry list (e2e repo vs sample repo) or explicit image names or something?

Reply:
I think future, it's out of scope for this issue.

Reply:
Agreed; I was just mentioning it as a point of interest. Wasn't sure if you had thought about that too. It will become slightly less of an issue if we can get those upstream images minimized.